### PR TITLE
Matrix testing of TensorFlow versions

### PR DESCRIPTION
**Pull request by johnamcleod**

### CODE REVIEW

File: .github/workflows/quality-check.yaml [hunk 1]
```diff
@@ -22,9 +22,18 @@ jobs:
...
+            tensorflow_probability: "~=0.10"
+          - python-version: 3.6
+            tensorflow: "~=2.3"
+            tensorflow_probability: "~=0.11"
           - python-version: 3.6
             tensorflow: "~=2.4"
             tensorflow_probability: "~=0.12"
+          - python-version: 3.6
+            tensorflow: "~=2.5"
+            tensorflow_probability: "~=0.12"
```
Comment (johnamcleod) [inline .github/workflows/quality-check.yaml]: 0.13.0 is available as a release candidate. When it is released this can be changed.

**Review by vdutor — APPROVED** [review]

File: gpflux/experiment_support/ci_utils.py [hunk 2]
```diff
@@ -38,21 +38,3 @@ def notebook_niter(n: int, test_n: int = 2) -> int:
...
-    return range(notebook_niter(n, test_n))
-
-
-def notebook_list(lst: list, test_n: int = 2) -> list:
-    """
-    Return a subset of the length specified by ``test_n`` from a list ``lst`` if the code
-    is running on CI machines (see :func:`is_continuous_integration`),
-    otherwise return the complete list ``lst``.
-    """
-    return lst[:test_n] if is_continuous_integration() else lst
```
Comment (vdutor) [inline gpflux/experiment_support/ci_utils.py:58]: why are we deleting these function? I appreciate that they are currently not being used, but they may be useful in the future when we have more notebooks?

File: tests/gpflux/layers/test_dedup_trackable_layer.py [hunk 3]
```diff
@@ -100,6 +101,7 @@ def _size_q_mu(num_inducing, output_dim):
 }
 
 
+@pytest.mark.skipif(Version(tf.__version__) <= Version("2.3"), reason="Do not run with TF-2.2")
```
Comment (vdutor) [inline tests/gpflux/layers/test_dedup_trackable_layer.py]: It looks like the "reason" and "if-statement" don't match. 
```suggestion
@pytest.mark.skipif(Version(tf.__version__) < Version("2.3"), reason="Do not run with TF-2.2")
```
or
```suggestion
@pytest.mark.skipif(Version(tf.__version__) <= Version("2.3"), reason="Do not run with TF-2.2 and TF-2.3")
```

File: tests/gpflux/layers/test_dedup_trackable_layer.py [hunk 4]
```diff
@@ -109,6 +111,9 @@ def test_count_weights(model):
     assert count_params(model) == int(sum(_MODEL_PARAMS_AND_SIZE.values()))
 
 
+@pytest.mark.skipif(
+    Version(tf.__version__) <= Version("2.4"), reason="Do not run with TF-2.2 or TF-2.3"
+)
```
Comment (vdutor) [inline tests/gpflux/layers/test_dedup_trackable_layer.py:116]: same as above.

File: tests/gpflux/layers/test_dedup_trackable_layer.py [hunk 5]
```diff
@@ -125,10 +130,14 @@ def test_parameter_names(model):
     assert set(_MODEL_PARAMS_AND_SIZE.keys()).issubset(set(params.keys()))
 
 
+@pytest.mark.skipif(Version(tf.__version__) <= Version("2.3"), reason="Do not run with TF-2.2")
```
Comment (vdutor) [inline tests/gpflux/layers/test_dedup_trackable_layer.py:133]: same as above.

File: tests/gpflux/layers/test_dedup_trackable_layer.py [hunk 6]
```diff
@@ -125,10 +130,14 @@ def test_parameter_names(model):
...
 def test_number_of_parameters(model):
     assert len(_MODEL_PARAMS_AND_SIZE) == len(model.trainable_weights)
 
 
+@pytest.mark.skipif(
+    Version(tf.__version__) <= Version("2.4"), reason="Do not run with TF-2.2 or TF-2.3"
+)
```
Comment (vdutor) [inline tests/gpflux/layers/test_dedup_trackable_layer.py:140]: same as above.

File: tests/gpflux/optimization/test_keras_natgrad.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+import numpy as np
+import pytest
+import tensorflow as tf
+
+import gpflow
+
+import gpflux
+from gpflux.layers.likelihood_layer import LikelihoodOutputs
+
+tf.keras.backend.set_floatx("float64")
```
Comment (vdutor) [inline tests/gpflux/optimization/test_keras_natgrad.py:28]: Do you know if this statement will affect the state for tests in other files?

File: tests/integration/test_svgp_equivalence.py [hunk 8]
```diff
@@ -245,6 +246,7 @@ def optimization_step():
         optimization_step()
 
 
+@pytest.mark.skipif(Version(tf.__version__) <= Version("2.3"), reason="Do not run with TF-2.2")
```
Comment (vdutor) [inline tests/integration/test_svgp_equivalence.py:249]: Please double-check:
```suggestion
@pytest.mark.skipif(Version(tf.__version__) < Version("2.3"), reason="Do not run with TF-2.2")
```

File: tests/integration/test_svgp_equivalence.py [hunk 9]
```diff
@@ -274,6 +276,9 @@ def test_svgp_equivalence_with_sldgp(svgp_fitter, sldgp_fitter, maxiter=20):
     ],
 )
 def test_svgp_equivalence_with_keras_sequential(svgp_fitter, keras_fitter, tol_kw, maxiter=10):
+    if svgp_fitter == fit_natgrad and Version(tf.__version__) <= Version("2.3"):
+        pytest.skip("Do not run with TF-2.2")
+
```
Comment (vdutor) [inline tests/integration/test_svgp_equivalence.py:281]: same as above.

File: tests/integration/test_svgp_equivalence.py [hunk 10]
```diff
@@ -329,6 +334,9 @@ def run_gpflow_svgp(data, optimizer, maxiter):
     ["natgrad", "adam", "scipy", "keras_adam", "keras_natgrad"],
 )
 def test_run_gpflux_sldgp(optimizer):
+    if optimizer == "keras_natgrad" and Version(tf.__version__) <= Version("2.3"):
+        pytest.skip("Do not run with TF-2.2")
+
```
Comment (vdutor) [inline tests/integration/test_svgp_equivalence.py:339]: same as above

File: gpflux/experiment_support/ci_utils.py (hunk 2 above)
Comment (johnamcleod) [reply gpflux/experiment_support/ci_utils.py:58]: They were not used, and they are not tested so they were dragging the average code coverage down. I can restore them and add a unit test to game the code coverage value.

**Comment by vdutor** [comment]
@johnamcleod, in light of #48, can this PR be closed?

**Comment by johnamcleod** [comment]
> @johnamcleod, in light of #48, can this PR be closed?

Yes

**Pull request closed without merging**